### ISSUE TITLE
taking upper triangular of "-inf" matrix results in nan values

### ISSUE THREAD

**Issue by wqlfff** [pytorch/pytorch#128429]
### 🐛 Describe the bug

```python
import torch
mask = torch.full((10, 10), float("-inf")).to("mps")
print(torch.triu(mask, diagonal=1))
```

### Versions

PyTorch version: 2.4.0.dev20240607
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.5 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.1.0.2.5)
CMake version: version 3.29.0
Libc version: N/A

Python version: 3.10.13 | packaged by conda-forge | (main, Oct 26 2023, 18:09:17) [Clang 16.0.6 ] (64-bit runtime)
Python platform: macOS-14.5-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] torch==2.4.0.dev20240607
[pip3] torchaudio==2.4.0.dev20240607
[pip3] torchvision==0.19.0.dev20240607
[conda] numpy                     1.22.0                   pypi_0    pypi
[conda] pytorch-lightning         2.2.4                    pypi_0    pypi
[conda] pytorch-metric-learning   2.5.0                    pypi_0    pypi
[conda] torch                     2.1.0                    pypi_0    pypi
[conda] torch-audiomentations     0.11.1                   pypi_0    pypi
[conda] torch-complex             0.4.3                    pypi_0    pypi
[conda] torch-pitch-shift         1.2.4                    pypi_0    pypi
[conda] torchaudio                2.1.0                    pypi_0    pypi
[conda] torchdata                 0.7.0                    pypi_0    pypi
[conda] torchmetrics              1.4.0.post0              pypi_0    pypi
[conda] torchtext                 0.16.0                   pypi_0    pypi
[conda] torchvision               0.16.0                   pypi_0    pypi

cc @kulinseth @albanD @malfet @DenisVieriu97 @jhavukainen

**Comment by jhavukainen** [comment]
@wqlfff Thanks for reporting the issue! Created a fix for this at: https://github.com/pytorch/pytorch/pull/128575

I'll update here once it's merged so you can test the nightly build to verify it solves the issue for you.
